# Make sure to submit your answers in the colab form field.
## Do not edit the code written in the form. We will be using the form for evaluating the answers.

In [1]:
import pandas as pd
import numpy as np

# Preprocessing

Part 1: Data preprocessing

DiabetesLevel (Target): if the person has diabetes

0 = no diabetes \\
1 = prediabetes \\
2 = diabetes \\
BloodPressure: Blood pressure of the patient

Cholestrol: cholesterol level of the patient

LastCholCheckupDate: date of date the patient had cholestrol checked last

Smoker: if the patient smokes regulraly

No = patient doesn't smoke or smokes seldomly
Yes = patient smokes regularly
Protein: 'Yes' if the patient consumes sufficient amount of protein reguarly in his diet, 'No' otherwise

Fats: 'Yes' if the patient consumes sufficient amount of fats reguarly in his diet, 'No' otherwise

Alchohol: 'Yes' if the patient consumes alchohol reguarly, 'No' otherwise

PhyHlth: For how many days during the past 30 days was your physical health NOT good (i.e. the patient might have had any physical illness and injury)? It is in days, scale will be between 0-30.

Sex: Gender of the patient,

'F = female
'M = male
Age: age of the patient in years.

Income: yearly income of the patient.

## Preprocessing Q1

In [2]:
#Code Here
df = pd.read_csv("PreprocessingDataset4.csv")

In [3]:
df.head(3)

DiabetesLevel  BloodPressure  Cholesterol LastCholCheckupDate Smoker  \
0            0.0            127          273          2020-05-29     No   
1            0.0            157          180          2021-06-23     No   
2            0.0            116          198          2022-01-19    Yes   

  Protein Fats Alchohol  PhysHlth  Age   Income Sex  
0     Yes  Yes      Yes       0.0  8.0  1160000   F  
1      No  Yes       No       0.0  9.0  1938200   M  
2      No  Yes       No       0.0  5.0  1503900   M

## Preprocessing Q2

How many samples are there in the dataset?


In [4]:
#Code Here
df.shape

(47276, 12)

## Preprocessing Q3

How many patients are suffering from either diabeties or prediabetes (i.e. include both)?

7838

7839

7831

7832

7833

None of these


In [7]:
#Code Here
df.query("(DiabetesLevel == 1) |  (DiabetesLevel == 2)").shape

(7838, 12)

In [5]:
df.head(3)

DiabetesLevel  BloodPressure  Cholesterol LastCholCheckupDate Smoker  \
0            0.0            127          273          2020-05-29     No   
1            0.0            157          180          2021-06-23     No   
2            0.0            116          198          2022-01-19    Yes   

  Protein Fats Alchohol  PhysHlth  Age   Income Sex  
0     Yes  Yes      Yes       0.0  8.0  1160000   F  
1      No  Yes       No       0.0  9.0  1938200   M  
2      No  Yes       No       0.0  5.0  1503900   M

## Preprocessing Q4


Consider following statistic: Number of days a patient reported he was phycially healthy in the last 30 days. What is the average of this statistic across all the patients.

Write your answer correct up to three decimal places.

In [13]:
df.PhysHlth.unique()

array([ 0.,  2.,  5., 30.,  3.,  1., 15., 23.,  4.,  7., 10., 14., 12.,
       20., 25.,  6., 21., 18.,  8., 28., 26., 16., 29., 17., 27., 22.,
        9., 13., 11., 19., 24.])

In [21]:
#Code Here

# PhyHlth: For how many days during the past 30 days was your
# physical health NOT good (i.e. the patient might have had
# any physical illness and injury)?
# It is in days, scale will be between 0-30.

# df.query("PhysHlth == 0")

# df.groupby('PhysHlth').agg('count').sort_values('DiabetesLevel')

(30 - df.PhysHlth).mean()

26.173618749471192

In [23]:
30 - df.PhysHlth.mean()

26.17361874947119

## Preprocessing Q5

How many female patients consume sufficient amount of protein or sufficient amount of fats or sufficient amount of both?

22337

22410

22294

22416

22439

None of these


In [25]:
df.head(2)

DiabetesLevel  BloodPressure  Cholesterol LastCholCheckupDate Smoker  \
0            0.0            127          273          2020-05-29     No   
1            0.0            157          180          2021-06-23     No   

  Protein Fats Alchohol  PhysHlth  Age   Income Sex  
0     Yes  Yes      Yes       0.0  8.0  1160000   F  
1      No  Yes       No       0.0  9.0  1938200   M

In [30]:
#Code Here
df.query("Sex == 'F'").query("(Protein == 'Yes') | (Fats == 'Yes')").shape

(22416, 12)

## Preprocessing Q6

Which income group has highest number of diabetes (exclude prediabetes)?

Income group:

Group 1 to 5:

1 = income  <  Rs. 3,00,000 \\
2 = Rs. 3,00,000  ≤  income  <  Rs. 6,00,000 \\
3 = Rs. 6,00,000  ≤  income  <  Rs. 9,00,000 \\
4 = Rs. 9,00,000  ≤  income  <  Rs. 15,00,000 \
5 = Rs. 15,00,000  ≤  income \\

1

2

3

4

5

Can not be determined from the data

There is a tie between two or more income groups

In [32]:
df.sample(3)

DiabetesLevel  BloodPressure  Cholesterol LastCholCheckupDate Smoker  \
17481            0.0            136          220          2019-06-06    Yes   
3952             0.0            119          228          2021-11-17    Yes   
30783            0.0            143          261          2021-05-07    Yes   

      Protein Fats Alchohol  PhysHlth   Age   Income Sex  
17481     Yes  Yes      Yes       0.0   9.0  1645000   M  
3952      Yes  Yes      Yes       0.0  10.0  1681200   F  
30783      No   No       No       0.0   5.0   323000   M

In [37]:
#Code Here
gp1 = df.query("Income < 300000")
gp2 = df.query("300000<= Income < 600000")
gp3 = df.query("600000<= Income < 900000")
gp4 = df.query("900000<= Income < 1500000")
gp5 = df.query("1500000 <= Income")

In [42]:
print(f"Number of diabetic in gp1: {(gp1.DiabetesLevel == 2).sum()}")
print(f"Number of diabetic in gp2: {(gp2.DiabetesLevel == 2).sum()}")
print(f"Number of diabetic in gp3: {(gp3.DiabetesLevel == 2).sum()}")
print(f"Number of diabetic in gp4: {(gp4.DiabetesLevel == 2).sum()}")
print(f"Number of diabetic in gp5: {(gp5.DiabetesLevel == 2).sum()}")

Number of diabetic in gp1: 473
Number of diabetic in gp2: 616
Number of diabetic in gp3: 1097
Number of diabetic in gp4: 3035
Number of diabetic in gp5: 1493


## Preprocessing Q7

What is the total number of missing values in the dataset?

19

14

11

18

13

None of these

In [43]:
#Code Here

df.isnull().sum()

DiabetesLevel           0
BloodPressure           0
Cholesterol             0
LastCholCheckupDate     0
Smoker                  0
Protein                 0
Fats                    0
Alchohol                0
PhysHlth                0
Age                    18
Income                  0
Sex                     0
dtype: int64

## Preprocessing Q8


How many patients had their last cholestrol checkup more than 5 years ago from today (i.e. 17 Dec 2023)?

1706

1759

1843

1771

1746

None of these

In [45]:
temp = df.sample(3)

In [53]:
temp.LastCholCheckupDate = pd.to_datetime(temp.LastCholCheckupDate)

In [62]:
import datetime

In [64]:
today = datetime.datetime(2023, 12, 17)

In [70]:
temp = (today - temp.LastCholCheckupDate)

In [73]:
(temp.dt.total_seconds()/60)/60/24/365

31518    4.164384
27261    4.178082
13599    4.082192
Name: LastCholCheckupDate, dtype: float64

In [76]:
#Code Here
df.LastCholCheckupDate = pd.to_datetime(df.LastCholCheckupDate)
today = datetime.datetime(2023, 12, 17)
q8 = ((((today - df.LastCholCheckupDate).dt.total_seconds()/60)/60)/24)/365

In [79]:
(q8 > 5).sum()

1771

## Preprocessing Q9

Which of the following categories have highest diabetes (not prediabetes) patients?

Male patients with income group 4 and don't smoke

Female patients with income group 5 and don't smoke

Male patients with income group 4 and smoke

Female with income group 4 and smoke

Can't be determined from the data

In [81]:
gp4.head(2)

DiabetesLevel  BloodPressure  Cholesterol LastCholCheckupDate Smoker  \
0            0.0            127          273          2020-05-29     No   
4            0.0            118          189          2023-04-08    Yes   

  Protein Fats Alchohol  PhysHlth   Age   Income Sex  
0     Yes  Yes      Yes       0.0   8.0  1160000   F  
4      No   No       No       5.0  10.0  1317800   M

In [88]:
#Code Here
op1 = gp4.query("(Sex == 'M') & (Smoker == 'No') ").query("DiabetesLevel == 2").shape[0]
op2 = gp5.query("(Sex == 'F') & (Smoker == 'No') ").query("DiabetesLevel == 2").shape[0]
op3 = gp4.query("(Sex == 'M') & (Smoker == 'Yes') ").query("DiabetesLevel == 2").shape[0]
op4 = gp4.query("(Sex == 'F') & (Smoker == 'Yes') ").query("DiabetesLevel == 2").shape[0]

print(f"option 1: {op1}")
print(f"option 2: {op2}")
print(f"option 3: {op3}")
print(f"option 4: {op4}")

option 1: 745
option 2: 310
option 3: 859
option 4: 569


## Preprocessing Q10

Median blood pressure of diabetic patients (exclude prediabetics)?

Note: Enter your answer correct upto 3 decimal place without rounding up or off.

In [91]:
#Code Here
df.query("DiabetesLevel == 2").BloodPressure.median()

118.0

## Preprocessing Q11

Apply Pre processing

Divide the data into training and test sets

Keep 30% of the data as test set.
Use random_state as 0
DiabetesLevel is the target, rest of the columsn are the features.
Apply stratified train test split according to labels in the dataset.
Hint: look for the documentation of the usual function that divides the data into training and test datasets.

What is the number of samples in the training set?

In [92]:
#Code Here
from sklearn.model_selection import train_test_split

X = df.drop("DiabetesLevel", axis=1)
y = df.DiabetesLevel

In [108]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0,stratify=y, test_size=0.3)

In [110]:
X_train.shape

(33093, 11)

## Preprocessing Q12

Apply following preprocessing steps on the training set:

Drop LastCholCheckupDate column

BloodPressure: min max scaling

Cholesterol: min max scaling

Ordinally encode Smoker

Ordinally encode Protein

Ordinally encode Fats

Ordinally encode Alchohol

PhysHlth: min max scaling

Age: impute with mean value then min max scaling

Income: min max scaling

One hot encode Sex column.

Hint: use column transformer, Then use this pipeline to transform the test set

What is the mean of the transformed test data (features only)?

In [116]:
X_train.columns

Index(['BloodPressure', 'Cholesterol', 'Smoker', 'Protein', 'Fats', 'Alchohol',
       'PhysHlth', 'Age', 'Income', 'Sex'],
      dtype='object')

In [118]:
min_max_attribs = ['BloodPressure', 'Cholesterol', 'PhysHlth', 'Age', 'Income']
ordinal_attribs = ['Smoker', 'Protein', 'Fats', 'Alchohol']
onehot_attribs = ['Sex']

In [119]:
#Code Here
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder, OrdinalEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

In [115]:
# X_train.drop('LastCholCheckupDate', axis=1, inplace=True)
# X_test.drop('LastCholCheckupDate', axis=1, inplace=True)

In [138]:
X_train.Age.replace(np.nan, X_train.Age.mean(), inplace=True)
X_test.Age.replace(np.nan, X_test.Age.mean(), inplace=True)

In [140]:
X_train.Age.isnull().sum()

0

In [141]:
X_test.Age.isnull().sum()

0

In [160]:
col_trans = ColumnTransformer([
    ('min_max', MinMaxScaler(), min_max_attribs),
    ('ordinal', OrdinalEncoder(), ordinal_attribs),
    ('onehot', OneHotEncoder(), onehot_attribs)
], remainder='passthrough', verbose_feature_names_out=False)

pipe = Pipeline([
    ('col_trans', col_trans)
])

In [161]:
X_train_transformed = pd.DataFrame(pipe.fit_transform(X_train, y_train), columns = col_trans.get_feature_names_out())
X_test_transformed = pd.DataFrame(pipe.transform(X_test), columns = col_trans.get_feature_names_out())

In [162]:
X_train.shape

(33093, 10)

In [165]:
X_train_transformed.head(3)

BloodPressure  Cholesterol  PhysHlth       Age    Income  Smoker  Protein  \
0       0.772727     0.828283       0.0  0.250000  0.577524     1.0      0.0   
1       0.886364     0.808081       0.0  0.416667  0.742439     0.0      1.0   
2       0.045455     0.595960       0.0  0.833333  0.400320     0.0      1.0   

   Fats  Alchohol  Sex_F  Sex_M  
0   1.0       0.0    1.0    0.0  
1   1.0       1.0    1.0    0.0  
2   1.0       0.0    1.0    0.0

In [171]:
X_test_transformed.mean().mean()


0.4577089566131182

## Preprocessing Q13

Consider diabetics and preiabetics, both as diabetic patients. Set the label for both of them as 1 in train and test label vectors. What is the count of diabetic samples in the training set after this transformation?



In [175]:
#Code Here
y_train.replace(2, 1, inplace=True)
y_test.replace(2, 1, inplace=True)

In [180]:
(y_train == 1).sum()

5487

#Model Building

In [296]:
df = pd.read_csv('modelBuilding3.csv')

In [299]:
df.tail()

0         1         2    3         4    5    6    7    8    9  \
47271  0.583333  0.045455  0.434343  0.0  0.575892  1.0  1.0  1.0  0.0  1.0   
47272  0.083333  0.227273  0.464646  0.0  0.440095  0.0  1.0  1.0  0.0  1.0   
47273  0.583333  0.068182  0.030303  0.0  0.560499  1.0  0.0  0.0  0.0  1.0   
47274  0.083333  0.386364  0.979798  0.0  0.921847  0.0  0.0  0.0  0.0  0.0   
47275  0.666667  0.295455  0.878788  0.0  0.629699  1.0  1.0  1.0  0.0  1.0   

        10   11  
47271  0.0  0.0  
47272  0.0  0.0  
47273  0.0  1.0  
47274  1.0  0.0  
47275  0.0  0.0

## Model Building Q1

In [300]:
#Code Here

print("modelBuilding3.csv")

modelBuilding3.csv


## Model Building Q2

Split the dataset into train dataset and test dataset in the following manner
	 Use train_test_split to split the dataset into train and test dataset with test size equal to 0.3(i.e.30%) and random_state equal to 42. Let other parameters have default values.

	Columns except the last column should be the feature matrix (X_train or X_test)

	Last column will be the label vector. It has only two labels, 1 for diabetes, 0 otherwise.

	The first row of the file has column identifiers (i.e. column names)
	The file does not have any index column.

Train KNN model on training set with  K=1,3,5,7 .

For what value of  K , the model has best accuracy score on test data?

1

3

5

7


There is a tie between two or more values of K

In [301]:
df.head(2)

0         1         2    3         4    5    6    7    8    9   10  \
0  0.333333  0.840909  0.595960  0.0  0.965377  0.0  1.0  1.0  0.0  1.0  0.0   
1  0.750000  0.454545  0.828283  0.0  0.769104  0.0  1.0  1.0  1.0  0.0  1.0   

    11  
0  0.0  
1  0.0

In [302]:
#Code Here
X = df.iloc[:, :-1]
y = df.iloc[:,-1]
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size=0.3)

In [303]:
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsClassifier

# [1,3,5,7]
knn1 = KNeighborsClassifier(n_neighbors=1)
knn3 = KNeighborsClassifier(n_neighbors=3)
knn5 = KNeighborsClassifier(n_neighbors=5)
knn7 = KNeighborsClassifier(n_neighbors=7)

In [304]:
knn1.fit(X_train, y_train)
knn3.fit(X_train, y_train)
knn5.fit(X_train, y_train)
knn7.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=7)

In [305]:
print(f"knn1 score: {knn1.score(X_test, y_test)}")
print(f"knn3 score: {knn3.score(X_test, y_test)}")
print(f"knn5 score: {knn5.score(X_test, y_test)}")
print(f"knn7 score: {knn7.score(X_test, y_test)}")

knn1 score: 0.7608404427836142
knn3 score: 0.8016639639004441
knn5 score: 0.81266304731016
knn7 score: 0.8206303320877106


## Model Building Q3

```
Use SGDClassifier on the training dataset (X_train and y_train) to train the model.
Use the following parameters:
	 log_loss:is the loss function to be used
	apply lasso regularization,
	maximum number of iterations is 10
	constant learning rate of 0.01,
	regularization rate value is 0.01,
	Take random_state=0.
	Set warm_start as False Note : Please ignore the convergence warning.
  
  Using above model, calculate and write the correct value of accuracy for the test set.
```

In [306]:
#Code Here
from sklearn.linear_model import SGDClassifier

sgd = SGDClassifier(loss='log_loss', penalty='l1', max_iter=10,
                    learning_rate='constant',
                    eta0=0.01,alpha=0.01,
                    random_state=0, warm_start=False)

In [307]:
sgd.fit(X_train, y_train)

SGDClassifier(alpha=0.01, eta0=0.01, learning_rate='constant', loss='log_loss',
              max_iter=10, penalty='l1', random_state=0)

In [308]:
from sklearn.metrics import accuracy_score

accuracy_score(y_test, sgd.predict(X_test))

0.8350842557991962

In [310]:
sgd.score(X_test, y_test)

0.8350842557991962

## Model Building Q4


```
(Common Instructions for Q.4 to Q.8)
Take DecisionTreeClassifier(random_state = 0) estimator with GridSearchCV. Hyperparameter tuning to be done over the following parameters:
	 Criterion as 'entropy' or 'gini'
	 Splitter as 'random' or 'best'
	 Minimum number of samples per leaf as [6,8,10]
	 Maximum depth as [5,6, 7, 8]
	 Minimum mumber of samples per node to split as [16, 18, 20]
	 Use cross validation = 5
	 Set refit to true
 Use the best model from above hyper parameter turing process to answer following questions:

What is the depth of the best estimator?

```

In [311]:
#Code Here
from sklearn.tree import DecisionTreeClassifier

dtree = DecisionTreeClassifier(random_state=0)

params = {
    'criterion': ['entropy', 'gini'],
    'splitter': ['random', 'best'],
    'min_samples_leaf': [6,8,10],
    'max_depth': [5,6,7,8],
    'min_samples_split': [16,18,20]
}

grid = GridSearchCV(dtree, param_grid=params, cv=5, refit=True)

In [312]:
grid.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=0),
             param_grid={'criterion': ['entropy', 'gini'],
                         'max_depth': [5, 6, 7, 8],
                         'min_samples_leaf': [6, 8, 10],
                         'min_samples_split': [16, 18, 20],
                         'splitter': ['random', 'best']})

In [313]:
dtree_tuned = grid.best_estimator_

In [314]:
grid.best_params_

{'criterion': 'entropy',
 'max_depth': 6,
 'min_samples_leaf': 6,
 'min_samples_split': 18,
 'splitter': 'best'}

In [315]:
dtree_tuned.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=6, min_samples_leaf=6,
                       min_samples_split=18, random_state=0)

## Model Building Q5

What is the number of the nodes in the tree?

393

63

127

217

399

In [316]:
#Code Here

dtree_tuned.tree_.node_count

127

## Model Building Q6

Calculate the total number of non-diabetic patients classified as diabetic and diabetic patients as non-diabetic, by the best estimator for the test data.

2364

2327

2339

2359

2375

None of these

In [317]:
#Code Here
from sklearn.metrics import confusion_matrix

cm_dtree = confusion_matrix(y_test, dtree_tuned.predict(X_test) )


In [318]:
cm_dtree

array([[11679,   165],
       [ 2174,   165]])

In [320]:
2174+165

2339

## Model Building Q7

What fraction of diabetic patients in the test set has been correctly classified as such by the best estimator?


In [328]:
(y_test ==0).sum()

11844

In [321]:
#Code Here
cm_dtree

array([[11679,   165],
       [ 2174,   165]])

In [325]:
cm_dtree.sum()

14183

In [342]:
165/(165+2174)

0.0705429670799487

In [344]:
recall_score(y_test, dtree_tuned.predict(X_test))

0.0705429670799487

## Model Building Q8

What fraction of patients classifed as diabetics by the best estimator, are actually diabetics, in the test set?


In [329]:
#Code Here
cm_dtree

array([[11679,   165],
       [ 2174,   165]])

In [334]:
165/(165+165)

0.5

In [335]:
precision_score(y_test, dtree_tuned.predict(X_test))

0.5

## Model Building Q9


6 points
Train a support vector machine classifier on the training set with random_state = 0 and rest of the parameters as default values. How many support vectors are there for this model?

14552

14613

14529

14893

14618

None of these

In [345]:
#Code Here
from sklearn.svm import SVC

svm_clf = SVC(random_state=0)


In [346]:
svm_clf.fit(X_train, y_train)

SVC(random_state=0)

In [347]:
svm_clf.support_vectors_.shape

(14759, 11)

In [348]:
svm_clf.n_support_

array([9260, 5499], dtype=int32)

In [349]:
9260+5499

14759